man/systemd.exec.xml

```diff
@@ -1644,6 +1644,24 @@ StateDirectory=aaa/bbb ccc</programlisting>
         <xi:include href="version-info.xml" xpointer="v234"/></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>StateDirectoryQuota=</varname></term>
+        <term><varname>CacheDirectoryQuota=</varname></term>
+        <term><varname>LogsDirectoryQuota=</varname></term>
+
+        <listitem><para>Specifies the storage limits for the directories specified in <varname>StateDirectory=</varname>,
+        <varname>CacheDirectory=</varname>, or <varname>LogsDirectory=</varname> respectively, as a percentage (e.g. 10%).
+        Defaults to <constant>0%</constant>, in which case no storage limits will be set.</para>
```

Comment:
hmm, i must say, i find it surprising this doesn't allow absolute limits to be specified. this will be very surprising to users.

I think we really should support both.

Reply:
My thinking there was that with a percentage, you can apply it to both blocks and inodes, as they are both important as both can be used to exhaust resources for other services.
And percentages are easier to understand and reason about in abstract, as you don't really need to know what the size of the disk is in advance, you just need to know how many services fighting among each other for resources you have.

If we want to support absolute values, then we need to come out with a good interface for it, that includes inode numbers too, and I don't really know how it could look like.

Reply:
frankly, i'd ignore inode number accounting. To me this appears to be an out of date concept to control that. the kernel's memory mgmt folks kinda moved towards tracking memory objects as memory for resource tracking (which is why we get away with raising fd limits so ridiculously high these days). I am pretty sure that inodes are similar out-of-date there: since the byte limits cover inodes too, indirectly i see no reason for bothering with that for now.

Hence, i'd really just ignore the inode limits, and focus on the byte limits only. If one day someone shows up who can make a good convincing argument that inodes limits must be managable too we can revisit things, but otherwise I'd really just ignore them.

Reply:
Well that might or might not happen in the distant future, but _today_ inode numbers are most definitely a resource that need management, as you _can_ DOS a system, since they are capped at the filesystem superblock level (can you even change the max inode number without re-creating the FS? Not sure tbh).

I've literally just experienced that with udev consuming all inodes on /run/ a couple of months ago and breaking production machines... unrelated to quotas for sure, but shows that the inode limit is very much a problem to deal with today, so I really do want a solution for that. I am fine if we set it only with the percentage setting however, and avoid an explicit knob, I don't need that.

Reply:
To put it another way: we need the quota system in place in order to ensure one service cannot stop another from storing its state/config/etc. If I have a quota of 10M, but I can create 0-sized files until the filesystem inode cap is reached, then the requirements are not fulfilled.

So we could have the setting take either a percentage, or the usual K/M/G/etc. If it's an absolute size, only the block quota is set. If it's a percentage, both the block and inode quotas are set. Does that work?